kotlinx-coroutines-core/common/src/flow/operators/Share.kt

```diff
@@ -412,6 +412,29 @@ private class SubscribedSharedFlow<T>(
         sharedFlow.collect(SubscribedFlowCollector(collector, action))
 }
 
+/**
+ * Returns a flow that invokes the given [action] **after** this state flow starts to be collected
+ * (after the subscription is registered).
+ *
+ * The [action] is called before any value is emitted from the upstream
+ * flow to this subscription but after the subscription is established. It is guaranteed that all emissions to
+ * the upstream flow that happen inside or immediately after this `onSubscription` action will be
+ * collected by this subscription.
+ *
+ * The receiver of the [action] is [FlowCollector], so `onSubscription` can emit additional elements.
+ */
+public fun <T> StateFlow<T>.onSubscription(action: suspend FlowCollector<T>.() -> Unit): StateFlow<T> =
```

Comment:
Please also add tests for this. Good inspirations are `SharedFlowTest#testOnSubscription` and `SharedFlowTest#onSubscriptionThrows`, which can be added to `StateFlowTest`. `testOnSubscriptionWithException` is quite complex and doesn't highlight the basic concepts of `onSubscription`, even though it probably tests them well.